### DISCUSSION TITLE
Pickling instrument objects

### DISCUSSION THREAD

**Discussion by stennebroek** [microsoft/Qcodes#3481] [Q&A]
https://github.com/QCoDeS/Qcodes/blob/67fc2b8db8e153fc2085340db9388b39c2ee9405/qcodes/instrument/base.py#L411-L415

Hi, I've got a question: Why is it that Qcodes `Instrument`s cannot or should not be pickled? Is the functionality not implemented (yet) or is there a more fundamental reason why they cannot be pickled?

Thanks a lot in advance!

**Comment by jenshnielsen** [comment]
Instruments often contain connections to physical instruments. These are often not safe to pickle and restore or move to a different thread. That being said if we understand your use case better perhaps we can find a way to enable what you need

**Reply by stennebroek** [reply]
Right, I understand. I am trying to find an easy way to save and reload a class - let's call it `Experiment` -  that has `Instrument`s and `Device`s as attributes, along with several other attributes (that might be objects as well). I thought that creating a function that loops recursively over all attributes of `Experiment`, but that feels like it will be a quite involved, especially when reloading the data in a similar `Experiment` class instance, say a couple of weeks later.

I'm quite new to QCoDeS, and I know about the snapshot option, but is there any way to "reload" a snapshot to an `Instrument` object such that an experiment can be repeated with exactly the same settings?

Thanks again!
